### DISCUSSION TITLE
[Nx] Huge difference in performance on a large monorepo compared to `nx`

### DISCUSSION THREAD

**Discussion by arkus7** [lemonade-hq/traf#20] [General]
I want to test `traf` on a real project using Nx monorepo. Unfortunately, the time `traf` requires to find the affected changes is super long, it takes ~405 seconds (6 minutes, 45 seconds) to find affected changes whereas `nx print-affected` does the job within 2.5 seconds.

Are there any "gotchas" when it comes to repository/code structure that I should be aware of, possibly causing the timings to be so big?

Here's a benchmark done with [`hyperfine`](https://github.com/sharkdp/hyperfine) on 14" Apple MacBook 2021 M1 Pro (16GB RAM)

```
hyperfine "npx nx print-affected --base origin/master --json" "npx @traf/nx@latest affected --base or
igin/master --json=true"             
Benchmark 1: npx nx print-affected --base origin/master --json
  Time (mean ± σ):      1.926 s ±  0.134 s    [User: 1.443 s, System: 0.681 s]
  Range (min … max):    1.831 s …  2.246 s    10 runs
 
Benchmark 2: npx @traf/nx@latest affected --base origin/master --json=true
  Time (mean ± σ):     405.872 s ±  6.739 s    [User: 475.691 s, System: 40.423 s]
  Range (min … max):   400.071 s … 420.857 s    10 runs
 
Summary
  'npx nx print-affected --base origin/master --json' ran
  210.73 ± 15.09 times faster than 'npx @traf/nx@latest affected --base origin/master --json=true'
```

My monorepo consists of:
- 34 apps
- 37 libs

Can you guide me to where can be my issue? Unfortunately, I can't share access to the repository.

Wasn't sure whether to post it as an issue or discussion, chose the latter.

**Comment by EladBezalel** [comment]
Hey 👋🏼,

First of all, I think traf as a tool can be better at telling what can be improved,

In general I think 6 mins is a lot but still way better than ci runs for unnecessary libs/apps.

A quick win will be to make sure tsconfigs aren't including spec/storybook files

Other than that, I have some improvements in mind,
Parallelize work, better caching, general performance improvements.

I'm more than willing to look into it together, would you like to schedule a call so we can look over your usecase?
